### PR TITLE
Added instrumentor::phase struct API implementation

### PR DESCRIPTION
**Pull request by tapaswenipathak**
As there is existing implementation of `instrumentor` struct, I have created a `namespace` named `instrumentor` other than `detail` and kept `phase` struct inside it. Please let me know if you think the design should be different.

Fix https://github.com/BlueBrain/CoreNeuron/issues/158.

### CODE REVIEW

File: coreneuron/nrniv/main1.cpp [hunk 1]
```diff
@@ -399,9 +399,8 @@ extern "C" int run_solve_core(int argc, char** argv) {
     }
 
     // initializationa and loading functions moved to separate
-    Instrumentor::phase_begin("load-model");
+    Instrumentor::phase p("load-model");
     nrn_init_and_load_data(argc, argv, !configs.empty());
-    Instrumentor::phase_end("load-model");
```
Comment (pramodk) [inline coreneuron/nrniv/main1.cpp]: `load-model` should only measure `nrn_init_and_load_data`. If I am not mistaken, `p` will measure more than this here (whole block scope). Need to introduce a local block? 

```
{
    Instrumentor::phase p("load-model");
    nrn_init_and_load_data(argc, argv, !configs.empty());
}
```

**Comment by tapaswenipathak** [comment]
@pramodk ping, do you think I should add anything else in the PR? Please let me know.

**Comment by tapaswenipathak** [comment]
hi @pramodk, do you think I can enhance anything else in the PR? Please let me know.

File: coreneuron/nrniv/profiler_interface.h [hunk 2]
```diff
@@ -260,6 +260,17 @@ struct NullInstrumentor {
 
 }  // namespace detail
 
+namespace Instrumentor {
+    struct phase {
+        phase() {
+            detail::Instrumentor::phase_begin();
```
Comment (pramodk) [inline coreneuron/nrniv/profiler_interface.h]: where do you capture the argument? (if this is instantiated as `p("checkpoint"`)). As per original semantic it was supposed to call `phase_begin("checkpoint")`.

Comment (tapaswenipathak) [reply coreneuron/nrniv/profiler_interface.h]: Implemented 2e743e4.

**Comment by tapaswenipathak** [comment]
hi @pramodk, do you have any more comments?

**Comment by pramodk** [comment]
Sorry @tapaswenipathak for not much response here. Everything looks fine, just would like to test this locally with instrumentation enabled to make sure Caliper works as expected.

> Some of us are busy with open sourcing of the [nmodl tool](https://github.com/BlueBrain/nmodl) for this project. Will get back to you in next few days.

**Comment by tapaswenipathak** [comment]
> Sorry @tapaswenipathak for not much response here. Everything looks fine, just would like to test this locally with instrumentation enabled to make sure Caliper works as expected.

> > Some of us are busy with open sourcing of the nmodl tool for this project. Will get back to you in next few days.

Oh okay, np, seems like @tristan0x is unavailable, great if you can do a quick review for https://github.com/BlueBrain/CoreNeuron/pull/170, https://github.com/BlueBrain/CoreNeuron/pull/167 as well, thanks.

**Comment by pramodk** [comment]
@tapaswenipathak : I installed [Caliper](https://github.com/LLNL/Caliper) and then try to build this branch as:

```
→ cmake .. -DENABLE_CALIPER=ON -Dcaliper_DIR=/Users/kumbhar/path/caliper-1.9.1/share/cmake/caliper
→ make 
...
[ 26%] Building CXX object coreneuron/CMakeFiles/coreneuron.dir/nrniv/main1.cpp.o
In file included from /Users/kumbhar/workarena/repos/bbp/coreneuron/coreneuron/nrniv/main1.cpp:54:
/Users/kumbhar/workarena/repos/bbp/coreneuron/coreneuron/nrniv/profiler_interface.h:266:21: error: 'Instrumentor' is not a class, namespace, or enumeration
            detail::Instrumentor::phase_begin(name);
                    ^
/Users/kumbhar/workarena/repos/bbp/coreneuron/coreneuron/nrniv/profiler_interface.h:38:8: note: 'Instrumentor' declared here
struct Instrumentor {
       ^
/Users/kumbhar/workarena/repos/bbp/coreneuron/coreneuron/nrniv/profiler_interface.h:269:21: error: 'Instrumentor' is not a class, namespace, or enumeration
            detail::Instrumentor::phase_end();
                    ^
/Users/kumbhar/workarena/repos/bbp/coreneuron/coreneuron/nrniv/profiler_interface.h:38:8: note: 'Instrumentor' declared here
struct Instrumentor {
       ^
/Users/kumbhar/workarena/repos/bbp/coreneuron/coreneuron/nrniv/profiler_interface.h:271:6: error: expected ';' after struct
    }
     ^
     ;
/Users/kumbhar/workarena/repos/bbp/coreneuron/coreneuron/nrniv/profiler_interface.h:274:7: error: redefinition of 'Instrumentor' as different kind of symbol
using Instrumentor = detail::Instrumentor<
      ^
/Users/kumbhar/workarena/repos/bbp/coreneuron/coreneuron/nrniv/profiler_interface.h:263:11: note: previous definition is here
namespace Instrumentor {
          ^
/Users/kumbhar/workarena/repos/bbp/coreneuron/coreneuron/nrniv/main1.cpp:167:19: error: no member named 'stop_profile' in namespace 'coreneuron::Instrumentor'
    Instrumentor::stop_profile();
    ~~~~~~~~~~~~~~^
/Users/kumbhar/workarena/repos/bbp/coreneuron/coreneuron/nrniv/main1.cpp:383:19: error: no member named 'phase_begin' in namespace 'coreneuron::Instrumentor'
```

**Comment by pramodk** [comment]
I now enabled tests/builds in internal CI. It's not accessible from GitHub but errors are obvious if we build simply as:

```
cmake ..
make 

# gives

[ 16%] Building CXX object coreneuron/CMakeFiles/coreneuron.dir/nrniv/main1.cpp.o
In file included from /gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/CNEUR-168/coreneuron/coreneuron/nrniv/main1.cpp:54:0:
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/CNEUR-168/coreneuron/coreneuron/nrniv/profiler_interface.h:281:11: error: ‘namespace coreneuron::Instrumentor { }’ redeclared as different kind of symbol
 namespace Instrumentor {
           ^~~~~~~~~~~~
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/CNEUR-168/coreneuron/coreneuron/nrniv/profiler_interface.h:279:30: note: previous declaration ‘using Instrumentor = struct coreneuron::detail::Instrumentor<coreneuron::detail::NullInstrumentor>’
     detail::NullInstrumentor>;
                              ^
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/CNEUR-168/coreneuron/coreneuron/nrniv/profiler_interface.h: In constructor ‘coreneuron::Instrumentor::phase::phase(const char*)’:
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/CNEUR-168/coreneuron/coreneuron/nrniv/profiler_interface.h:284:21: error: ‘template<class ... TProfilerImpl> struct coreneuron::detail::Instrumentor’ used without template parameters
             detail::Instrumentor::phase_begin(name);
                     ^~~~~~~~~~~~
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/CNEUR-168/coreneuron/coreneuron/nrniv/profiler_interface.h: In destructor ‘coreneuron::Instrumentor::phase::~phase()’:
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/CNEUR-168/coreneuron/coreneuron/nrniv/profiler_interface.h:287:21: error: ‘template<class ... TProfilerImpl> struct coreneuron::detail::Instrumentor’ used without template parameters
             detail::Instrumentor::phase_end();
                     ^~~~~~~~~~~~
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/CNEUR-168/coreneuron/coreneuron/nrniv/main1.cpp: In function ‘int run_solve_core(int, char**)’:
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/CNEUR-168/coreneuron/coreneuron/nrniv/main1.cpp:403:9: error: ‘phase’ is not a member of ‘coreneuron::Instrumentor {aka coreneuron::detail::Instrumentor<coreneuron::detail::NullInstrumentor>}’
         Instrumentor::phase p("load-model");
         ^~~~~~~~~~~~
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/CNEUR-168/coreneuron/coreneuron/nrniv/main1.cpp:487:9: error: ‘phase’ is not a member of ‘coreneuron::Instrumentor {aka coreneuron::detail::Instrumentor<coreneuron::detail::NullInstrumentor>}’
         Instrumentor::phase p("output-spike");
         ^~~~~~~~~~~~
/gpfs/bbp.cscs.ch/project/proj20/pramod_scratch/CNEUR-168/coreneuron/coreneuron/nrniv/main1.cpp:492:9: error: ‘phase’ is not a member of ‘coreneuron::Instrumentor {aka coreneuron::detail::Instrumentor<coreneuron::detail::NullInstrumentor>}’
         Instrumentor::phase p("checkpoint");
         ^~~~~~~~~~~~
```

**Comment by pramodk** [comment]
@tapaswenipathak : I took liberty to push changes to your branch. Now it should build fine. If you could check if I have missed anything my changes? Otherwise looks fine to me.

**Comment by tapaswenipathak** [comment]
Thanks @pramodk, appreciate the effort.

**Pull request merged**